### PR TITLE
[RFC] Add ikev2 parser v01

### PR DESCRIPTION
**Pull request by chifflier**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

This PR adds support for a new protocol, IKEv2 (used for IPsec key exchange, RFC 7296).
- It follows the first messages exchanged on UDP port 500 to extract the handshake parameters. The client sends a list of proposals, the parameters are chosen by the server
- It is written in rust, and uses the crate `ipsec-parser` to decode the messages.
- It adds a logger for the handshake parameters (SPIs, exchange type, cryptographic parameters, notify messages, etc.) in JSON format
- It adds new rules for IKEv2 events: decoding errors, weak cryptographic proposals (deprecated encryption ciphers, small DH groups, etc.).

Example of JSON output (only "ikev2" key):
``
{"version_major":2,"version_minor":0,"exchange_type":34,"message_id":0,"init_spi":"729d50d0d8008dc1","resp_spi":"31713cfe052ad2fe","role":"responder","alg_enc":"ENCR_AES_GCM_16","alg_auth":"NONE","alg_prf":"PRF_HMAC_SHA2_256","alg_dh":"brainpoolP512r1","alg_esn":"NoESN","errors":0,"payload_list":"SecurityAssociation,KeyExchange,Nonce,Notify,Notify,CertificateRequest,Notify,Notify,NoNextPayload","notify_list":"NAT_DETECTION_SOURCE_IP,NAT_DETECTION_DESTINATION_IP,IKEV2_FRAGMENTATION_SUPPORTED,MULTIPLE_AUTH_SUPPORTED"}}
``

The following is planned for the future:
- when NAT-T is enabled, some key exchange messages are encapsulated and sent on port 4500/udp.  The same parser can be reused for this port, with some changes to ignore ESP data
- the list of deprecated/weak algorithm parameters could be a configuration/data file provided to the configuration. This would allow sharing rules with TLS, for ex.

### CODE REVIEW

**Comment by norg** [comment]
some documentation missing :)?

**Comment by chifflier** [comment]
I pushed a (small) update to documentation (given that there are only events, no new keywords yet). Please tell me if there is I need to add more things

File: rust/src/ikev2/ikev2.rs [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+    InvalidProposal,
+    UnknownProposal,
+}
+
+pub struct IKEV2State {
+    /// List of transactions for this session
+    transactions: Vec<IKEV2Transaction>,
+
+    /// Events counter
+    events: u16,
```
Comment (victorjulien) [inline rust/src/ikev2/ikev2.rs]: I think this is no longer needed.

File: rust/src/ikev2/ikev2.rs [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+    fn parse(&mut self, i: &[u8], direction: u8) -> i8 {
+        match parse_ikev2_header(i) {
+            IResult::Done(rem,ref hdr) => {
+                SCLogDebug!("parse_ikev2: {:?}",hdr);
+                if rem.len() == 0 && hdr.length == 28 {
+                    return 1;
+                }
+                // Rule 0: check version
+                if hdr.maj_ver != 2 || hdr.min_ver != 0 {
+                    SCLogInfo!("Unknown header version: {}.{}", hdr.maj_ver, hdr.min_ver);
```
Comment (victorjulien) [inline rust/src/ikev2/ikev2.rs]: this should be an event without a SCLogInfo

File: rust/src/ikev2/ikev2.rs [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+                }
+                if hdr.init_spi == 0 {
+                    SCLogInfo!("Malformed SPI - wrong length");
```
Comment (victorjulien) [inline rust/src/ikev2/ikev2.rs]: please turn this into SCLogDebug

File: rust/src/ikev2/ikev2.rs [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+                                // XXX TSr
+                                // XXX IDr
+                                _ => {
+                                    SCLogInfo!("Unknown payload content {:?}", payload.content);
+                                },
+                            }
+                            self.connection_state = self.connection_state.advance(payload);
+                        };
+                    },
+                    e @ _ => SCLogInfo!("parse_ikev2_payload_with_type: {:?}",e),
```
Comment (victorjulien) [inline rust/src/ikev2/ikev2.rs]: what is this notation/syntax?

File: src/output-json-ikev2.c [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+} LogIKEv2LogThread;
+
+static int JsonIKEv2Logger(ThreadVars *tv, void *thread_data,
+    const Packet *p, Flow *f, void *state, void *tx, uint64_t tx_id)
+{
+    IKEV2Transaction *ikev2tx = tx;
+    LogIKEv2LogThread *thread = thread_data;
+    json_t *js, *ikev2js;
+
+    js = CreateJSONHeader((Packet *)p, 0, "ikev2");
```
Comment (victorjulien) [inline src/output-json-ikev2.c]: this has been updated in master. See https://github.com/OISF/suricata/commit/710c7b821fcb919dfaa816f45a8b342acee622f4

**Comment by victorjulien** [comment]
The output example in nicer formatting:

```json
{
  "version_major": 2,
  "version_minor": 0,
  "exchange_type": 34,
  "message_id": 0,
  "init_spi": "729d50d0d8008dc1",
  "resp_spi": "31713cfe052ad2fe",
  "role": "responder",
  "alg_enc": "ENCR_AES_GCM_16",
  "alg_auth": "NONE",
  "alg_prf": "PRF_HMAC_SHA2_256",
  "alg_dh": "brainpoolP512r1",
  "alg_esn": "NoESN",
  "errors": 0,
  "payload_list": "SecurityAssociation,KeyExchange,Nonce,Notify,Notify,CertificateRequest,Notify,Notify,NoNextPayload",
  "notify_list": "NAT_DETECTION_SOURCE_IP,NAT_DETECTION_DESTINATION_IP,IKEV2_FRAGMENTATION_SUPPORTED,MULTIPLE_AUTH_SUPPORTED"
}
```

I think it makes sense to turn those lists into json arrays.

cc: @jasonish any thoughts on the output format?

File: rust/src/ikev2/ikev2.rs (hunk 1 above)
Comment (chifflier) [reply rust/src/ikev2/ikev2.rs]: Ok, removed

File: rust/src/ikev2/ikev2.rs (hunk 2 above)
Comment (chifflier) [reply rust/src/ikev2/ikev2.rs]: Changed to a `MalformedData` event

File: rust/src/ikev2/ikev2.rs (hunk 3 above)
Comment (chifflier) [reply rust/src/ikev2/ikev2.rs]: Log removed as well

File: rust/src/ikev2/ikev2.rs (hunk 4 above)
Comment (chifflier) [reply rust/src/ikev2/ikev2.rs]: This is a way to name parameters in a match arm (https://doc.rust-lang.org/book/first-edition/patterns.html#bindings).
That said, naming `_` is not useful, this is equivalent to
```rust
match ... {
    e => SCLogInfo!(...)
```
I removed the `@ _` part

File: src/output-json-ikev2.c (hunk 5 above)
Comment (chifflier) [reply src/output-json-ikev2.c]: OK, fixed

**Comment by chifflier** [comment]
@victorjulien Are the JSON arrays ordered ? The lists here represent the payloads and notify messages, in order of appearance in the message.

**Comment by victorjulien** [comment]
I think they are yes.

**Comment by chifflier** [comment]
Updated by #3311

**Pull request closed without merging**